// @duneanalytics/client-sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Paginator = exports.TableAPI = exports.CustomAPI = exports.ExecutionAPI = exports.QueryAPI = exports.DuneClient = void 0;
var api_1 = require("./api");
Object.defineProperty(exports, "DuneClient", { enumerable: true, get: function () { return api_1.DuneClient; } });
Object.defineProperty(exports, "QueryAPI", { enumerable: true, get: function () { return api_1.QueryAPI; } });
Object.defineProperty(exports, "ExecutionAPI", { enumerable: true, get: function () { return api_1.ExecutionAPI; } });
Object.defineProperty(exports, "CustomAPI", { enumerable: true, get: function () { return api_1.CustomAPI; } });
Object.defineProperty(exports, "TableAPI", { enumerable: true, get: function () { return api_1.TableAPI; } });
__exportStar(require("./types"), exports);
var paginator_1 = require("./paginator");
Object.defineProperty(exports, "Paginator", { enumerable: true, get: function () { return paginator_1.Paginator; } });
